File: helfomat-api/src/main/java/de/helfenkannjeder/helfomat/api/picture/ResizeImageService.java
package de.helfenkannjeder.helfomat.api.picture;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.file.Path;

/**
 * @author Valentin Zickner
 */
@Service
public class ResizeImageService {

    private final Logger LOGGER = LoggerFactory.getLogger(ResizeImageService.class);

    public void resize(Path input, Path output, final Integer expectedWidth, final Integer expectedHeight) throws IOException {
        LOGGER.debug("Scale '" + input + "' to size " + expectedWidth + "x" + expectedHeight + " with new name '" + output + "' ");
        BufferedImage image = ImageIO.read(input.toFile());

        int width = image.getWidth();
        int height = image.getHeight();

        if (isNecessaryToScaleHeight(expectedWidth, expectedHeight, image)) {
            width = expectedWidth;
            height = (expectedWidth * image.getHeight()) / image.getWidth();
        } else if (isNecessaryToScaleWidth(expectedWidth, expectedHeight, image)) {
            height = expectedHeight;
            width = (expectedHeight * image.getWidth()) / image.getHeight();
        }

        image = scaleImage(image, width, height);

        if (isNecessaryToCropHeight(expectedHeight, height)) {
            image = cropImage(image, 0, (height - expectedHeight) / 2, expectedWidth, expectedHeight);
        } else if (isNecessaryToCropWidth(expectedWidth, width)) {
            image = cropImage(image, (width - expectedWidth) / 2, 0, expectedWidth, expectedHeight);
        }

        ImageIO.write(image, "PNG", output.toFile());
    }

    private static BufferedImage scaleImage(BufferedImage image, int width, int height) {
        Image tempImage = image.getScaledInstance(width, height, Image.SCALE_SMOOTH);
        image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        Graphics graphics = image.getGraphics();
        graphics.drawImage(tempImage, 0, 0, null);
        graphics.dispose();
        return image;
    }

    private static BufferedImage cropImage(BufferedImage image, int x, int y, int width, int height) {
        image = image.getSubimage(x, y, width, height);
        return image;
    }

    private static boolean isNecessaryToScaleHeight(Integer expectedWidth, Integer expectedHeight, BufferedImage image) {
        return (
            expectedWidth != null && expectedHeight == null
        ) || (
            expectedHeight != null
                && expectedWidth != null
                && image.getWidth() * expectedHeight < expectedWidth * image.getHeight()
        );
    }

    private static boolean isNecessaryToScaleWidth(Integer expectedWidth, Integer expectedHeight, BufferedImage image) {
        return (
            expectedWidth == null && expectedHeight != null
        ) || (
            expectedWidth != null
                && expectedHeight != null
                && image.getWidth() * expectedHeight > expectedWidth * image.getHeight()
        );
    }

    private static boolean isNecessaryToCropHeight(Integer expectedHeight, int height) {
        return expectedHeight != null && height != expectedHeight;
    }

    private static boolean isNecessaryToCropWidth(Integer expectedWidth, int width) {
        return expectedWidth != null && width != expectedWidth;
    }

}

File: helfomat-api/src/main/kotlin/de/helfenkannjeder/helfomat/api/picture/ResizeImageService.kt
package de.helfenkannjeder.helfomat.api.picture

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.awt.Image
import java.awt.image.BufferedImage
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import javax.imageio.ImageIO


/**
 * @author Valentin Zickner
 */
@Service
open class ResizeImageService {

    open fun resize(input: InputStream, expectedWidth: Int?, expectedHeight: Int?, contentType: String?): Pair<InputStream, Long> {
        LOGGER.debug("Scale '$input' to size ${expectedWidth ?: "-"}x${expectedHeight ?: "-"}")
        var image = ImageIO.read(input)
        var width = image.width
        var height = image.height
        if (expectedWidth != null && expectedHeight == null || expectedHeight != null && expectedWidth != null && isNecessaryToScaleHeight(expectedWidth, expectedHeight, image)) {
            width = expectedWidth
            height = expectedWidth * image.height / image.width
        } else if (expectedWidth == null && expectedHeight != null || expectedWidth != null && expectedHeight != null && isNecessaryToScaleWidth(expectedWidth, expectedHeight, image)) {
            height = expectedHeight
            width = expectedHeight * image.width / image.height
        }
        val resultContentType = getContentType(contentType)
        image = scaleImage(image, width, height, resultContentType)
        if (expectedHeight != null && height != expectedHeight) {
            image = cropImage(image, 0, (height - expectedHeight) / 2, expectedWidth ?: width, expectedHeight)
        } else if (expectedWidth != null && width != expectedWidth) {
            image = cropImage(image, (width - expectedWidth) / 2, 0, expectedWidth, expectedHeight ?: height)
        }
        val imageWriters = ImageIO.getImageWritersByMIMEType(resultContentType)
        val hasNext = imageWriters.iterator().hasNext()
        if (!hasNext) {
            throw IllegalStateException("Failed to find image writer for '$contentType'")
        }
        val imageWriter = imageWriters.next()
        val byteArrayOutputStream = ByteArrayOutputStream()
        imageWriter.output = ImageIO.createImageOutputStream(byteArrayOutputStream)
        imageWriter.write(image)
        val inputStream = ByteArrayInputStream(byteArrayOutputStream.toByteArray())
        return Pair(inputStream, byteArrayOutputStream.size().toLong())
    }

    private fun getContentType(contentType: String?) =
        when (contentType) {
            "image/gif" -> "image/png"
            "image/png" -> "image/png"
            else -> "image/jpeg"
        }

    private fun scaleImage(inputImage: BufferedImage, width: Int, height: Int, resultContentType: String): BufferedImage {
        var image = inputImage
        val tempImage = image.getScaledInstance(width, height, Image.SCALE_SMOOTH)
        var type = BufferedImage.TYPE_INT_RGB
        if (hasAlphaChannel(resultContentType)) {
            type = BufferedImage.TYPE_INT_ARGB
        }
        image = BufferedImage(width, height, type)
        val graphics = image.graphics
        graphics.drawImage(tempImage, 0, 0, null)
        graphics.dispose()
        return image
    }

    private fun hasAlphaChannel(resultContentType: String) = resultContentType != "image/jpeg"

    private fun cropImage(inputImage: BufferedImage, x: Int, y: Int, width: Int, height: Int): BufferedImage {
        var image = inputImage
        image = image.getSubimage(x, y, width, height)
        return image
    }

    private fun isNecessaryToScaleHeight(expectedWidth: Int, expectedHeight: Int, image: BufferedImage): Boolean {
        return image.width * expectedHeight < expectedWidth * image.height
    }

    private fun isNecessaryToScaleWidth(expectedWidth: Int, expectedHeight: Int, image: BufferedImage): Boolean {
        return image.width * expectedHeight > expectedWidth * image.height
    }

    companion object {
        private var LOGGER = LoggerFactory.getLogger(ResizeImageService::class.java)

    }

}